packages/web-app-files/src/components/SideBar/Shares/Collaborators/RoleDropdown.vue

```diff
@@ -144,16 +153,42 @@ export default {
     customPermissionsRole() {
       return PeopleShareRoles.custom(this.resource.isFolder)
     },
+    resourceIsSharable() {
+      return this.allowSharePermission && this.resource.canShare()
+    },
+    share() {
+      const userShares = this.sharesTree[this.resource.path]?.filter((s) =>
+        ShareTypes.containsAnyValue(ShareTypes.individuals, [s.shareType])
+      )
+
+      return userShares?.length ? userShares[0] : undefined
+    },
+    allowCustomSharing() {
+      return this.capabilities?.files_sharing?.allow_custom
+    },
     availableRoles() {
       if (this.resourceIsSpace) {
         return SpacePeopleShareRoles.list()
       }
-      return PeopleShareRoles.list(
-        this.resource.isFolder,
-        this.capabilities?.files_sharing?.allow_custom !== false
-      )
+
+      if (this.resource.isReceivedShare() && this.resourceIsSharable && this.share) {
+        return PeopleShareRoles.listByBitmask(
+          parseInt(this.share.permissions),
+          this.resource.isFolder,
+          this.allowSharePermission,
+          this.allowCustomSharing !== false
+        )
+      }
+
+      return PeopleShareRoles.list(this.resource.isFolder, this.allowCustomSharing !== false)
     },
     availablePermissions() {
+      if (this.resource.isReceivedShare() && this.resourceIsSharable && this.share) {
+        return this.customPermissionsRole.permissionsByBitmask(
```

Comment:
Shouldn't this be done via `SharePermissions.bitmaskToPermissions`?

Reply:
Ahh of course, thx!